### ISSUE TITLE
JetStream WorkQueue invalid NumDelivered values when consumer fails

### ISSUE THREAD

**Issue by toserk** [nats-io/nats-server#5346]
### Observed behavior

I have a WorkQueue to process long running jobs (AckWait set to 30 sec and InProgress used). When consumer fails without Acking or Nacking the message (OutOfMemory or power failure), NumDelivered starts to report invalid values. Sometimes (rare) it get stuck and I get infinite redeliveries.

### Expected behavior

NumDelivered incremented when consumer fails

### Server and client version

nats-server: v2.10.14
.NET NATS.Client: 1.1.4

### Host environment

OS: Windows 10 x64
NET version: 3.1
Server started from win x64 binary, no docker

### Steps to reproduce

Sample code to reproduce:

```
static void Main(string[] args)
{
    using var connection = new ConnectionFactory().CreateConnection("nats://localhost");
    var jetStream = connection.CreateJetStreamContext();
    
    // Main process
    if (!ulong.TryParse(args.LastOrDefault(), out var expectedNum))
    {
        var jsm = connection.CreateJetStreamManagementContext();
        var streamInfo = jsm.GetStreams().FirstOrDefault(s => s.Config.Name == "jobs");
        if (streamInfo != null)
            jsm.DeleteStream("jobs");

        jsm.AddStream(StreamConfiguration.Builder().WithName("jobs").WithRetentionPolicy(RetentionPolicy.WorkQueue)
            .WithMaxAge(Duration.OfMillis(TimeSpan.FromDays(7).Milliseconds)).WithSubjects("my.queues.*").Build());
        jetStream.Publish("my.queues.test", new byte[] { 1, 2, 3 });
        foreach (var procNum in Enumerable.Range(1, 10))
        {
            // Start child process
            var psi = new ProcessStartInfo(Process.GetCurrentProcess().MainModule?.FileName,
                procNum.ToString());
            var process = new Process();
            process.StartInfo = psi;
            process.Start();
            Thread.Sleep(15000);
            
            // Emulate failure
            process.Kill();
        }

        Console.ReadKey();
        return;
    }

    // Child process
    Console.WriteLine(expectedNum);
    var consumerConfig = ConsumerConfiguration.Builder().WithAckWait(30000).WithMaxDeliver(15)
        .WithInactiveThreshold(60000).Build();
    var subOptions = PullSubscribeOptions.Builder().WithName("testconsumer").WithConfiguration(consumerConfig).Build();
    using var sub = jetStream.PullSubscribe("my.queues.test", subOptions);

    Msg msg;
    while (true)
    {
        var received = sub.Fetch(1, 1000);
        if (received.Count == 0)
            continue;

        msg = received[0];
        break;
    }

    if (msg.MetaData.NumDelivered != expectedNum)
        Console.WriteLine($"FAIL (was {msg.MetaData.NumDelivered})");

    while (true)
    {
        Thread.Sleep(10000);
        msg.InProgress();
    }
}
```

Program output:
1
2
3
FAIL (was 1)
4
5
FAIL (was 3)
6
7
FAIL (was 4)
8
9
FAIL (was 5)
10

**Comment by derekcollison** [comment]
Not sure if applicable but this PR is in main and will be pulled into release/v2.10.15 shortly.

https://github.com/nats-io/nats-server/pull/5338

**Comment by toserk** [comment]
> Not sure if applicable but this PR is in main and will be pulled into release/v2.10.15 shortly.
> 
> #5338

I don't think it is related, in my case message IS delivered but NumDelivery not updated or updated to invalid value

**Comment by mtmk** [comment]
> looking to repro...

I was able to reproduce similar output (thanks @toserk for the example) both with 2.10.14 and main. looking at the details now.

**Comment by mtmk** [comment]
@toserk it looks like server is doing the right thing. add this line to see number of deliveries incrementing sequentially:
```csharp
// ...
        msg = received[0];
        break;
    }

    Console.WriteLine($"[{expectedNum}] NumDelivered: {msg.MetaData.NumDelivered}"); // <-- add this line

    if (msg.MetaData.NumDelivered != expectedNum)
        Console.WriteLine($"FAIL (was {msg.MetaData.NumDelivered})");

    while (true)
// ...
```

it's because `ack_wait` is set to 30s and processes are killed after 15s not every process is getting a redelivery (roughly every other is getting one). If you set `ack_wait` to 10s, for example, you should see each process getting a redelivery. (of course, one shouldn't rely on timings to synchronize workers.)

> Sometimes (rare) it get stuck and I get infinite redeliveries.

Not sure about this one. afaik you shouldn't see a redelivery after reaching `max_deliver` (_The number of times a message will be redelivered to consumers if not acknowledged in time_)

**Comment by toserk** [comment]
My bad. In case of this example this is indeed my mistake. But my actual code don't have this mistake (since it don't have process.Kill logic).
I will try to update example later.

**Comment by toserk** [comment]
So, the first part of my problem (abouot arbitrary values in DeliveryNum) is my mistake, server behaves as intended.
The second part (about DeliveryNum get stuck on 1 and infinite redeliveries) I am able reproduce in actual app, and not able to reproduce in test project (and I copied all my nats related code as is to the test project et this point)
The only difference I found so far is what when I call PullSubscribe in actual app with catch all exceptions enabled, I get a lot of FormatExceptions inside Nats client without useful stacktrace, and I dont get them in test project.
Will try to build NATS.Client and debug it tomorrow. It looks like client in actual app don't communicate to server the fact what message was delivered for some reason.
![photo_2024-04-24_21-57-04](https://github.com/nats-io/nats-server/assets/22008064/44c92f1c-52ce-4cb7-a37a-a75d879abc19)
